### DISCUSSION TITLE
eslint migration to flat config issues

### DISCUSSION THREAD

**Discussion by adam-marshall** [nrwl/nx#26600] [General]
Is there any useful documentation about the process of migrating eslint to the flat configuration file?

I am just completing a migration from Nx 15 to 19 (one version at a time) and hitting lots of issues with eslint.

I had a setup with `eslint-plugin-jest` which migrated to:

```
  ...compat.config({ plugins: ['jest'] }).map(config => ({
    ...config,
    files: ['**/*.spec.ts'],
    rules: {
      'jest/expect-expect': ['error', { assertFunctionNames: ['expect', 'expect*', '*.expect*'] }],
      'jest/no-identical-title': 'error',
      'jest/valid-expect': 'error',
    },
  })),
```

But now my `.spec.ts` files (newly generated and existing) are subjected to other non-Jest lint rules and are failing e.g. `error  'describe' is not defined    no-undef`

Similarly I get `parserOptions.project` errors now where I didn't previously, i.e. https://github.com/nrwl/nx/discussions/8312. 

The syntax seems a little uglier than the documentation, there is a lot of spreading going on and `FlatCompat` being used.

So question: has anyone undergone this process, with a significant existing eslint configuration, and can share any advice or tips?

**Comment by adam-marshall** [comment]
Just noticed in https://nx.dev/recipes/tips-n-tricks/flat-config#correctness-and-best-practices it states

> Correctness and best practices
The purpose of this generator is to create a flat config that works the same way as the original JSON config did. Depending on the complexity of your original config, it may be using the FlatCompat utility to provide a compatibility wrapper around parts of the original config. You can improve those by following the [official migration guide](https://eslint.org/docs/latest/use/configure/migration-guide).

This answers the part of my question about the syntax.

If I generate a new, clean workspace, then migrate to flat config - its all fine.

So this gives me a starting point to troubleshoot further.

**Reply by adam-marshall** [reply]
ok that was easier than I thought...

adding `env: { jest: true }` did the trick, i.e.

```
...compat.config({ env: { jest: true }, plugins: ['jest'] }).map(config => ({
    ...config,
    files: ['**/*.spec.ts'],
    rules: {
      'jest/expect-expect': ['error', { assertFunctionNames: ['expect', 'expect*', '*.expect*'] }],
      'jest/no-identical-title': 'error',
      'jest/valid-expect': 'error',
    },
  })),
```

**Comment by adam-marshall** [comment]
With the `parserOptions.project` problems, which also affected Webstorm, I ended up going back to pre-flat config.

**Comment by ak274** [comment]
Is there any plan to support flat config by default while creating new  workspaces? 
Because flat config is now default and recommended in ESLint Docs.

https://eslint.org/blog/2024/04/eslint-v9.0.0-released/#flat-config-is-now-the-default-and-has-some-changes
